File: moira/src/bridge/clotho.ts
import { invoke } from '@tauri-apps/api/core'

import type {
  ForgeLocalBuildRequestPayload,
  InstallPublishedReleaseRequestPayload,
  KnownLocalBuildRegistrationPayload,
  LaunchTargetRefPayload,
  LaunchTargetSummaryPayload,
  ProfileDocumentPayload,
  ProfileDocumentSummaryPayload,
  ProfileRefPayload,
  PublishedReleaseSummaryPayload,
  SaveProfileDocumentRequestPayload,
} from '@/bridge/contracts/clotho'

export async function listLaunchTargets(): Promise<LaunchTargetSummaryPayload[]> {
  return invoke<LaunchTargetSummaryPayload[]>('list_launch_targets')
}

export async function registerKnownLocalBuild(
  registration: KnownLocalBuildRegistrationPayload,
): Promise<LaunchTargetRefPayload> {
  return invoke<LaunchTargetRefPayload>('register_known_local_build', { registration })
}

export async function forgeLocalBuild(
  request: ForgeLocalBuildRequestPayload,
): Promise<LaunchTargetRefPayload> {
  return invoke<LaunchTargetRefPayload>('forge_local_build', { request })
}

export async function listPublishedReleases(): Promise<PublishedReleaseSummaryPayload[]> {
  return invoke<PublishedReleaseSummaryPayload[]>('list_published_releases')
}

export async function installPublishedRelease(
  request: InstallPublishedReleaseRequestPayload,
): Promise<LaunchTargetRefPayload> {
  return invoke<LaunchTargetRefPayload>('install_published_release', { request })
}

export async function listProfileDocuments(): Promise<ProfileDocumentSummaryPayload[]> {
  return invoke<ProfileDocumentSummaryPayload[]>('list_profile_documents')
}

export async function loadProfileDocument(profile: ProfileRefPayload): Promise<ProfileDocumentPayload> {
  return invoke<ProfileDocumentPayload>('load_profile_document', { profile })
}

export async function saveProfileDocument(
  request: SaveProfileDocumentRequestPayload,
): Promise<ProfileDocumentPayload> {
  return invoke<ProfileDocumentPayload>('save_profile_document', { request })
}

File: moira/src-tauri/src/app/commands/clotho.rs
use tauri::State;

use crate::{
    app::state::AppState,
    clotho::model::{
        ForgeLocalBuildRequest, InstallPublishedReleaseRequest, KnownLocalBuildRegistration,
        LaunchTargetRef, LaunchTargetSummary, PreparedWakeInput, ProfileDocument,
        ProfileDocumentSummary, ProfileRef, PublishedReleaseSummary, SaveProfileDocumentRequest,
        WakeInputRequest,
    },
};

#[tauri::command]
pub async fn register_known_local_build(
    registration: KnownLocalBuildRegistration,
    state: State<'_, AppState>,
) -> Result<LaunchTargetRef, String> {
    state.clotho.register_known_local_build(registration)
}

#[tauri::command]
pub async fn prepare_wake_input(
    request: WakeInputRequest,
    state: State<'_, AppState>,
) -> Result<PreparedWakeInput, String> {
    state.clotho.prepare_wake_input(&request)
}

#[tauri::command]
pub async fn forge_local_build(
    request: ForgeLocalBuildRequest,
    state: State<'_, AppState>,
) -> Result<LaunchTargetRef, String> {
    state.clotho.forge_local_build(request).await
}

#[tauri::command]
pub async fn list_launch_targets(
    state: State<'_, AppState>,
) -> Result<Vec<LaunchTargetSummary>, String> {
    state.clotho.list_launch_targets()
}

#[tauri::command]
pub async fn list_published_releases(
    state: State<'_, AppState>,
) -> Result<Vec<PublishedReleaseSummary>, String> {
    state.clotho.list_published_releases().await
}

#[tauri::command]
pub async fn install_published_release(
    request: InstallPublishedReleaseRequest,
    state: State<'_, AppState>,
) -> Result<LaunchTargetRef, String> {
    state.clotho.install_published_release(request).await
}

#[tauri::command]
pub async fn list_profile_documents(
    state: State<'_, AppState>,
) -> Result<Vec<ProfileDocumentSummary>, String> {
    state.clotho.list_profile_documents()
}

#[tauri::command]
pub async fn load_profile_document(
    profile: ProfileRef,
    state: State<'_, AppState>,
) -> Result<ProfileDocument, String> {
    state.clotho.load_profile_document(&profile)
}

#[tauri::command]
pub async fn save_profile_document(
    request: SaveProfileDocumentRequest,
    state: State<'_, AppState>,
) -> Result<ProfileDocument, String> {
    state.clotho.save_profile_document(request)
}
